src/super_gradients/training/models/segmentation_models/unet/unet_decoder.py

```diff
@@ -0,0 +1,182 @@
+from typing import List, Type, Union
+from abc import ABC, abstractmethod
+from enum import Enum
+
+import torch.nn as nn
+import torch
+
+from super_gradients.training.utils.module_utils import ConvBNReLU, make_upsample_module
+
+
+class AbstractUpFuseBlock(nn.Module, ABC):
+    """
+    Abstract class for upsample and fuse UNet decoder building block.
+    """
+    def __init__(self, in_channels: int, skip_channels: int, out_channels: int, **kwargs):
+        """
+        :param in_channels: num_channels of the feature map to be upsample.
+        :param skip_channels: num_channels of the skip feature map from higher resolution.
+        :param out_channels: num_channels of the output features.
+        """
+        super().__init__()
+
+    @abstractmethod
+    def forward(self, x, skip):
+        raise NotImplementedError()
+
+
+class UpFactorBlock(AbstractUpFuseBlock):
+    """
+    Ignore Skip features, simply apply upsampling and ConvBNRelu layers.
+    """
+    def __init__(self,
+                 in_channels: int,
+                 skip_channels: int,
+                 out_channels: int,
+                 up_factor: int,
+                 mode: str,
+                 num_repeats: int,
+                 **kwargs
+                 ):
+        super().__init__(in_channels=in_channels, skip_channels=0, out_channels=out_channels)
+        self.up_path = make_upsample_module(scale_factor=up_factor, upsample_mode=mode, align_corners=False)
+
+        self.last_convs = nn.Sequential(
+            ConvBNReLU(in_channels, out_channels, kernel_size=3, padding=1, bias=False),
+            nn.Sequential(*[ConvBNReLU(out_channels, out_channels, kernel_size=3, padding=1, bias=False) for _ in range(num_repeats - 1)]),
+        )
+
+    def forward(self, x, skip):
+        x = self.up_path(x)
+        return self.last_convs(x)
+
+
+class UpCatBlock(AbstractUpFuseBlock):
+    """
+    Fuse features with concatenation and followed Convolutions.
+    """
+    def __init__(self,
+                 in_channels: int,
+                 skip_channels: int,
+                 out_channels: int,
+                 up_factor: int,
+                 mode: str,
+                 num_repeats: int,
+                 **kwargs
+                 ):
+        super().__init__(in_channels=in_channels, skip_channels=skip_channels, out_channels=out_channels)
+        self.up_path = make_upsample_module(scale_factor=up_factor, upsample_mode=mode, align_corners=False)
+        self.last_convs = nn.Sequential(
+            ConvBNReLU(in_channels + skip_channels, out_channels, kernel_size=3, padding=1, bias=False),
+            nn.Sequential(*[ConvBNReLU(out_channels, out_channels, kernel_size=3, padding=1, bias=False) for _ in range(num_repeats - 1)]),
+        )
+
+    def forward(self, x, skip):
+        x = self.up_path(x)
+        x = torch.cat([x, skip], dim=1)
+        return self.last_convs(x)
+
+
+class UpBlockType(Enum):
```

Comment:
please follow the logic of factories:
- define a dictionary of possible UpFuseBlocks (possibly, move them all to another file)
- define a factory 
- use a decorator over Decoder.__init__() to create a UpFuseBlock from a string

Reply:
Since the params for the blocks are dynamic parameters that depends on other layers in the models, such as in_channels, the factory is not required to initiate an object of the block class, only required to return a class type from string. For this matter I created a factory class called `TypeFactory` which can be initiated from enum or from a dictionary. [f33631f](https://github.com/Deci-AI/super-gradients/pull/401/commits/f33631f226d77345abd9db0c38b369015efc6e17)

Reply:
ok. very nice